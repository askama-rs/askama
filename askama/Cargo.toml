[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
