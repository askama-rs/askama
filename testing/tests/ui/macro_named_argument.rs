use askama::Template;

#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2) -%}
{{ param1 }} {{ param2 }}
{%- endmacro -%}

{%- call thrice(param1=2, param3=3) -%}", ext = "html")]
struct InvalidNamedArg;

#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2) -%}
{{ param1 }} {{ param2 }}
{%- endmacro -%}

{%- call thrice(param1=2, param1=3) -%}", ext = "html")]
struct InvalidNamedArg2;

// Ensures that filters can't have named arguments.
#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2) -%}
{{ param1 }} {{ param2 }}
{%- endmacro -%}

{%- call thrice(3, param1=2)|filter(param1=12) -%}", ext = "html")]
struct InvalidNamedArg3;

// Ensures that named arguments can only be passed last.
#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2) -%}
{{ param1 }} {{ param2 }}
{%- endmacro -%}
{%- call thrice(param1=2, 3) -%}", ext = "html")]
struct InvalidNamedArg4;

// Ensures that named arguments can't be used for arguments before them.
#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2) -%}
{{ param1 }} {{ param2 }}
{%- endmacro -%}
{%- call thrice(3, param1=2) -%}", ext = "html")]
struct InvalidNamedArg5;

#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2) -%}
{%- endmacro -%}
{%- call thrice() -%}", ext = "html")]
struct MissingArgs;

#[derive(Template)]
#[template(source = "{%- macro thrice(param1, param2=1) -%}
{%- endmacro -%}
{%- call thrice() -%}", ext = "html")]
struct MissingArgs2;

fn main() {
}
