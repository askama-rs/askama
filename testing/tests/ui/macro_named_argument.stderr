error: missing argument when calling macro `thrice`: `param2`
 --> InvalidNamedArg.html:5:4
       "call thrice(param1=2, param3=3) -%}{%- endcall -%}"
 --> tests/ui/macro_named_argument.rs:4:21
  |
4 |   #[template(source = "{%- macro thrice(param1, param2) -%}
  |  _____________________^
5 | | {{ param1 }} {{ param2 }}
6 | | {%- endmacro -%}
7 | |
8 | | {%- call thrice(param1=2, param3=3) -%}{%- endcall -%}", ext = "html")]
  | |_______________________________________________________^

error: named argument `param1` was passed more than once
 --> <source attribute>:5:26
       "param1=3) -%}{%- endcall -%}"
  --> tests/ui/macro_named_argument.rs:12:21
   |
12 |   #[template(source = "{%- macro thrice(param1, param2) -%}
   |  _____________________^
13 | | {{ param1 }} {{ param2 }}
14 | | {%- endmacro -%}
15 | |
16 | | {%- call thrice(param1=2, param1=3) -%}{%- endcall -%}", ext = "html")]
   | |_______________________________________________________^

error: failed to parse template source
 --> <source attribute>:5:28
       "|filter(param1=12) -%}{%- endcall -%}"
  --> tests/ui/macro_named_argument.rs:21:21
   |
21 |   #[template(source = "{%- macro thrice(param1, param2) -%}
   |  _____________________^
22 | | {{ param1 }} {{ param2 }}
23 | | {%- endmacro -%}
24 | |
25 | | {%- call thrice(3, param1=2)|filter(param1=12) -%}{%- endcall -%}", ext = "html")]
   | |__________________________________________________________________^

error: named arguments must always be passed last
 --> <source attribute>:4:26
       "3) -%}{%- endcall -%}"
  --> tests/ui/macro_named_argument.rs:30:21
   |
30 |   #[template(source = "{%- macro thrice(param1, param2) -%}
   |  _____________________^
31 | | {{ param1 }} {{ param2 }}
32 | | {%- endmacro -%}
33 | | {%- call thrice(param1=2, 3) -%}{%- endcall -%}", ext = "html")]
   | |________________________________________________^

error: argument `param1` was passed more than once when calling macro `thrice`
 --> InvalidNamedArg5.html:4:19
       "param1=2) -%}{%- endcall -%}"
  --> tests/ui/macro_named_argument.rs:38:21
   |
38 |   #[template(source = "{%- macro thrice(param1, param2) -%}
   |  _____________________^
39 | | {{ param1 }} {{ param2 }}
40 | | {%- endmacro -%}
41 | | {%- call thrice(3, param1=2) -%}{%- endcall -%}", ext = "html")]
   | |________________________________________________^

error: missing arguments when calling macro `thrice`: `param1` and `param2`
 --> MissingArgs.html:3:4
       "call thrice() -%}{%- endcall -%}"
  --> tests/ui/macro_named_argument.rs:45:21
   |
45 |   #[template(source = "{%- macro thrice(param1, param2) -%}
   |  _____________________^
46 | | {%- endmacro -%}
47 | | {%- call thrice() -%}{%- endcall -%}", ext = "html")]
   | |_____________________________________^

error: missing argument when calling macro `thrice`: `param1`
 --> MissingArgs2.html:3:4
       "call thrice() -%}{%- endcall -%}"
  --> tests/ui/macro_named_argument.rs:51:21
   |
51 |   #[template(source = "{%- macro thrice(param1, param2=1) -%}
   |  _____________________^
52 | | {%- endmacro -%}
53 | | {%- call thrice() -%}{%- endcall -%}", ext = "html")]
   | |_____________________________________^
